Signal/src/contact/SystemContactsFetcher.swift

```diff
@@ -6,6 +6,300 @@ import Foundation
+            let contactFetchRequest = CNContactFetchRequest(keysToFetch: self.allowedContactKeys)
+            try self.contactStore.enumerateContacts(with: contactFetchRequest) { (contact, _) -> Void in
+                systemContacts.append(contact)
+            }
+        } catch let error as NSError {
+            Logger.error("\(self.TAG) Failed to fetch contacts with error:\(error)")
+            assertionFailure()
+            return .error(error)
+        }
+
+        let contacts = systemContacts.map { Contact(systemContact: $0) }
+        return .success(contacts)
+    }
+}
+
+let kAddressBookContactyStoreDidChangeNotificationName = NSNotification.Name("AddressBookContactStoreAdapteeDidChange")
+/**
+ * System contact fetching compatible with iOS8
+ */
+class AddressBookContactStoreAdaptee: ContactStoreAdaptee {
+
+    let TAG = "[AddressBookContactStoreAdaptee]"
+
+    private var addressBook: ABAddressBook = ABAddressBookCreateWithOptions(nil, nil).takeRetainedValue()
+    private var changeHandler: (() -> Void)?
+    let supportsContactEditing = false
+
+    var authorizationStatus: ContactStoreAuthorizationStatus {
+        switch ABAddressBookGetAuthorizationStatus() {
+        case .notDetermined:
+            return .notDetermined
+        case .restricted:
+            return .restricted
+        case .denied:
+            return .denied
+        case .authorized:
+            return .authorized
+        }
+    }
+
+    @objc
+    func runChangeHandler() {
+        guard let changeHandler = self.changeHandler else {
+            Logger.error("\(TAG) trying to run change handler before it was registered")
+            assertionFailure()
+            return
+        }
+        changeHandler()
+    }
+
+    func startObservingChanges(changeHandler: @escaping () -> Void) {
+        // should only call once
+        assert(self.changeHandler == nil)
+        self.changeHandler = changeHandler
+
+        NotificationCenter.default.addObserver(self, selector: #selector(runChangeHandler), name: kAddressBookContactyStoreDidChangeNotificationName, object: nil)
+
+        let callback: ABExternalChangeCallback = { (_, _, _) in
+            // Ideally we'd just call the changeHandler here, but because this is a C style callback in swift, 
+            // we can't capture any state in the closure, so we use a notification as a trampoline
+            NotificationCenter.default.post(name: kAddressBookContactyStoreDidChangeNotificationName, object: nil)
+        }
+
+        ABAddressBookRegisterExternalChangeCallback(addressBook, callback, nil)
+    }
+
+    func requestAccess(completionHandler: @escaping (Bool, Error?) -> Void) {
+        ABAddressBookRequestAccessWithCompletion(addressBook, completionHandler)
+    }
+
+    func fetchContacts() -> Result<[Contact], Error> {
+        // Changes are not reflected unless we create a new address book
+        self.addressBook = ABAddressBookCreateWithOptions(nil, nil).takeRetainedValue()
```

Comment:
Yes, it looks redundant. I initially did not have this.

But then I saw that changes to the address book were not reflected in the app. I peaked at our old AB integration and saw we were doing something like this every time contacts were updated.

Hence the comment